server/app/report.go

```diff
@@ -0,0 +1,228 @@
+	"fmt"
+	"os"
+	"path"
+
+	"github.com/pkg/errors"
+
+	"github.com/mattermost/mattermost-plugin-user-survey/server/model"
+	"github.com/mattermost/mattermost-plugin-user-survey/server/utils"
+)
+
+const (
+	rawResponsePerPage = 500
+)
+
+func (a *UserSurveyApp) GenerateSurveyReport(userID, surveyID string) (*os.File, error) {
+	reportFilePath, err := a.generateSurveyReport(surveyID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "GenerateSurveyReport: userID: %s, surveyID: %s", userID, surveyID)
+	}
+
+	file, err := os.Open(reportFilePath)
+	if err != nil {
+		a.api.LogError("GenerateSurveyReport: failed to open survey report zip file for reading", "filePath", reportFilePath, "error", err.Error())
+		return nil, errors.Wrapf(err, "GenerateSurveyReport: failed to open survey report zip file for reading, filePath: %s", reportFilePath)
+	}
+
+	return file, nil
+}
+
+func (a *UserSurveyApp) generateSurveyReport(surveyID string) (string, error) {
+	survey, err := a.store.GetSurveysByID(surveyID)
+	if err != nil {
+		return "", errors.Wrapf(err, "generateSurveyReport: failed to get survey by ID, surveyID: %s", surveyID)
+	}
+
+	key := utils.NewID()
+
+	rawResponseCSVFilePath, err := a.generateRawResponseCSV(survey, key)
+	if err != nil {
+		return "", err
+	}
+
+	surveyMetadataFilePath, err := a.generateSurveyMetadataFile(survey, key)
+	if err != nil {
+		return "", err
+	}
+
+	zipPath := path.Join(os.TempDir(), "survey_report", key, "survey_report.zip")
+	files := []string{rawResponseCSVFilePath, surveyMetadataFilePath}
+	err = utils.CreateZip(zipPath, files)
+	if err != nil {
+		a.api.LogError("generateSurveyReport: failed to generate report zip file", "surveyID", surveyID, "error", err.Error())
+		return "", errors.Wrapf(err, "generateSurveyReport: failed to generate report zip file, surveyID: %s", surveyID)
+	}
+
+	// delete the unneeded files now
+	for _, filePath := range files {
+		a.api.LogDebug("generateSurveyReport: deleting file", "filePath", filePath)
+		err = os.Remove(filePath)
+		if err != nil {
+			// not a critical error, no need to break
+			a.api.LogWarn("generateSurveyReport: failed to delete file path", "filePath", filePath, "error", err.Error())
+		}
+	}
+
+	return zipPath, nil
+}
+
+func (a *UserSurveyApp) generateRawResponseCSV(survey *model.Survey, key string) (string, error) {
+	var lastResponseID string
+
+	headers := []string{"User ID", "Submitted At"}
+	for _, question := range survey.SurveyQuestions.Questions {
+		headers = append(headers, question.Text)
+	}
+
+	part := 0
+	for {
+		// get a page worth of results
+		data, err := a.store.GetAllResponses(survey.ID, lastResponseID, rawResponsePerPage)
+		if err != nil {
+			return "", err
+		}
+
+		if len(data) == 0 {
+			break
+		}
+
+		lastResponseID = data[len(data)-1].ID
+
+		// save them in a temp CSV
+		if err := a.saveTempCSVData(key, part, data, survey); err != nil {
+			return "", errors.Wrapf(err, "generateRawResponseCSV: surveyID: %s", survey.ID)
+		}
```

Comment:
I discussed this and Agniva and the recommendation was to store each part CSV on disk rather than in memory to avoid any accidental or ill-intentional out of memory issues.

Reply:
+1 on this - we did the same thing for the generated report for user admin reporting.

Reply:
@devinbinnie exactly. In fact I use your code as a reference for this.